nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        self.dropout = dropout
+        self.conv = torch.nn.Sequential(
+            OrderedDict(
+                [
+                    module
+                    for i in range(len(cnn_filters))
+                    for module in (
+                        ("transpose_bchw_{}".format(i + 1), Transpose([0, 3, 1, 2])),
+                        (
+                            "conv2d_{}".format(i + 1),
+                            torch.nn.Conv2d(
+                                in_channels=int(self.filter_size[i]),
+                                out_channels=int(self.filter_size[i + 1]),
+                                kernel_size=(3, 3),
+                                stride=(2, 2),
+                                padding=(1, 1),
+                                dilation=(1, 1),
+                                bias=True,
+                            ),
+                        ),
+                        ("transpose_bhwc_{}".format(i + 1), Transpose([0, 2, 3, 1])),
+                        ("relu_{}".format(i + 1), torch.nn.ReLU()),
+                        ("layer_norm_{}".format(i + 1), torch.nn.LayerNorm(self.filter_size[i + 1]),),
+                        ("dropout_{}".format(i + 1), torch.nn.Dropout(self.dropout)),
+                    )
+                ]
+            )
+        )
+
+        self.gru = torch.nn.GRU(input_size=cnn_filters[-1] * 2, hidden_size=gru_hidden, batch_first=True,)
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inputs_masks": NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "out": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs, inputs_masks):
+        inputs = inputs.transpose(1, 2)
+
+        inputs = inputs * inputs_masks
+        out = inputs.unsqueeze(3)
+        out = self.conv(out)
+        out = out.view(out.shape[0], out.shape[1], -1).contiguous()
+        self.gru.flatten_parameters()
+        _, out = self.gru(out)
+
+        return out.squeeze(0)
+
+
+class StyleAttention(NeuralModule):
+    def __init__(self, gru_hidden=128, gst_size=128, n_style_token=10, n_style_attn_head=4):
+        super(StyleAttention, self).__init__()
+        self.input_size = gru_hidden
+        self.output_size = gst_size
+        self.n_token = n_style_token
+        self.n_head = n_style_attn_head
+        self.token_size = self.output_size // self.n_head
+
+        self.tokens = torch.nn.Parameter(torch.FloatTensor(self.n_token, self.token_size))
+
+        self.q_linear = torch.nn.Linear(self.input_size, self.output_size)
+        self.k_linear = torch.nn.Linear(self.token_size, self.output_size)
+        self.v_linear = torch.nn.Linear(self.token_size, self.output_size)
+
+        self.tanh = torch.nn.Tanh()
+        self.softmax = torch.nn.Softmax(dim=2)
```

Comment:
Can we use https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html ?

Reply:
I think the parameters should be called `scale` instead of `temperature`. And It looks like `nn.MultiheadAttention` do the scaling inside in these two parts [line1](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5294) and [line2](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5348). Let me take a look at how to map my pre-trained weights to `nn.MultiheadAttention`. If it has some mismatch, I'll change this maybe for next PR.

Reply:
I change the implementation with `torch.nn.MultiheadAttention`.